Update rights csv function

bonniegee opened the pull request:
Fixes #170

bonniegee committed:
update rights csv function

helrond reviewed (commented):
While this reduces the number of lines of code, IMO it's slightly harder to understand, mostly because of the multiple `append` statements. Before doing anything else, though, I'd suggest making this method a bit more isolated and testable:
- move the actual writing to csv (`csvwriter.writerow`) up to the `create_rights_csv` method, and just return a list from this function. This function should then probably be renamed, something like `get_rights_row`.
- Add a test for this method, which asserts output on static input.

Once that's done, we can take another look at this method. Things I'd consider:
- Is all of the `get(foo, '')` really necessary? The rights statements get validated when bags are saved to Ursa Major, so we can know for sure what keys will be there, and which might not. Even if some keys might not reliably be there, it might be better to catch `KeyErrors` with a try/except loop here, rather than waiting for CSV validation.
- I think we determined that capitalization in `rights_basis` doesn't matter, so that `capitalize` could probably go away, which would simplify things.
- Similarly, I would check on the `restriction`/`grant_restriction` keys - I would bet that they are required and we could do something like `rights_granted.get('restriction', rights_granted['grant_restriction'])`

bonniegee commented:
>     * Is all of the `get(foo, '')` really necessary? The rights statements get validated when bags are saved to Ursa Major, so we can know for sure what keys will be there, and which might not. Even if some keys might not reliably be there, it might be better to catch `KeyErrors` with a try/except loop here, rather than waiting for CSV validation.

I think in order to reliably know whether a key is expected/required/optional, we need to incorporate checking for the rights basis. I'm not sure if the additionally complexity is worth it, but am willing to be convinced otherwise!

bonniegee committed:
split rights row functionary into dif functions

bonniegee commented:
This could be cleaned up more but I wanted to push up the approach I've been taking. Some notes:

* Unless I'm misunderstanding, I think we weren't creating rows if there wasn't an act? That's why I added that `len() == 0` check, but let me know if I'm missing something.
* I split getting the basis information and the grant/restriction information out into two different functions. This is causing some weirdness because the act fields are in the middle of the basis fields, but I'm hoping this approach is more readable? It also can lay the foundation for breaking out this functionality by rights basis.
* I know I need docstrings, just wanted to make sure I was heading in the right direction first...

bonniegee commented:
Oh, one more thing--my tests failed when I tried `rights_granted.get('restriction', rights_granted['grant_restriction'])`, which is why I stuck with the original approach.

helrond commented on sip_assembly/routines_helpers.py:
```diff
@@ -98,36 +96,56 @@ def setup_csv_file(filepath):
+            rights_row.append(os.path.join(bag_dir, file).lstrip('/'))
+            for basis_value in get_basis_fields(rights_statement):
+                rights_row.append(basis_value)
+            count = 10
+            for grant_value in get_grant_restriction(rights_granted):
+                rights_row.insert(count, grant_value)
+                count += 1
```
How about if `get_grant_restriction` returns a list (instead of a tuple) and then you just stick the two together using `+` or `extend`?

helrond replied:
Also, if `get_grant_restriction` always returns a list with the right number of empty strings, then you don't need this case splitting.

helrond replied:
Oh sorry, just realized that what I'm suggesting in the first comment won't work because the act fields are in the middle of the rights statement fields. I think you can use the list slicing method instead (and you could use a tuple if you want):
```
rights_row[10:10] = get_grant_restriction(rights_granted)
```

helrond commented on sip_assembly/routines_helpers.py:
```diff
@@ -98,36 +96,56 @@ def setup_csv_file(filepath):
+
+def get_basis_fields(rights_statement):
+    """docstring for get_basis_fields"""
+    basis_values = []
+    for field in ['rights_basis', 'status', 'determination_date', 'jurisdiction', 'start_date', 'end_date', 'terms', 'citation', 'note', 'doc_id_type', 'doc_id_value', 'doc_id_role']:
+        basis_values.append(rights_statement.get(field, ''))
+    return basis_values
```
you could do this a little more elegantly with list comprehension:
```
basis_fields = [...]
return [rights_statement.get(field, "") for field in basis_fields]

helrond commented on sip_assembly/routines_helpers.py:
```diff
@@ -68,13 +68,11 @@ def create_rights_csv(bag_path, rights_statements):
-                csvwriter)
+        for (dirpath, dirnames, filenames) in os.walk(os.path.join(bag_path, 'data', 'objects')):
+            for file in filenames:
+                rights_rows = get_rights_rows(dirpath.split(bag_path)[1], file, rights_statement)
```
Worth looking at this first argument - are we just passing "objects" because if so it might be more transparent to just pass that string rather than doing all this splitting and index referral business.

helrond reviewed (commented):
I think this is headed in the right direction. I have some thoughts on how you can simplify further...

bonniegee committed:
simplify basis field method

bonniegee commented:
Thanks, @helrond ! Making those changes now. I'm concerned about rights statements that do not have acts--I'm going to add a fixture that represents that.

bonniegee commented:
@helrond - Working on creating tests now, and I have half a mind to move all the csv-related helpers to a new file/class called like CSVCreator. Would this help with maintainability or is it overkill?

helrond commented:
Sure - I think that would be fine. As noted in #172 I think some of those helpers are (or could be) redundant of code in Asterism.

bonniegee committed:
update urlize

bonniegee committed:
refactor csv creation

bonniegee committed:
Merge branch 'issue-170' of https://github.com/RockefellerArchiveCenter/fornax into issue-170

bonniegee committed:
clean up error codes

bonniegee committed:
actually call validator

bonniegee committed:
insert blank columns in right place

bonniegee committed:
close csv file before validating

bonniegee committed:
close csv file, return message at end of run

bonniegee committed:
check for copyright status or status

and only test status value if basis is copyright

bonniegee committed:
only validate date fields if not empty

bonniegee committed:
clean up check restriction method

bonniegee committed:
add docstrings

bonniegee committed:
update csv creation fixtures

bonniegee committed:
add tests

helrond committed:
simplify logic

helrond committed:
tweak docstrings

helrond committed:
test all the rows

helrond reviewed (commented):
Cool, this looks good - I tweaked the logic a bit to move the for loop for the rights granted information down into the get_granted_fields method (and renamed the method as well). This allowed me to get rid of a bunch of lines in get_rights_rows which were doing mostly the same thing.

bonniegee commented:
Update--I deployed this branch on dev and it's producing* rights CSVs that fail with our current version of Archivematica, but will work with 1.13.

*Only in cases where there's no grant or restriction attached to a basis.

helrond commented:
Is there a way of producing rights CSVs that work on current Archivematica as well as 1.13?

bonniegee committed:
account for multiple am versions

bonniegee committed:
Merge branch 'issue-170' of https://github.com/RockefellerArchiveCenter/fornax into issue-170

bonniegee committed:
actually fix merge conflicts

helrond reviewed (approved):
Love. It.

This is really nice - I think setting `skip_no_act` attribute is a really nice way of making that logic super transparent.
